### DISCUSSION TITLE
Is it legal to use primitive values as unknown?

### DISCUSSION THREAD

**Discussion by yamarten** [bluesky-social/atproto#2615] [Clarifications and "Why?"] [answered]
[The lexicon validator](https://github.com/bluesky-social/atproto/blob/52a596320addd0111a8c347df3a9e85b4bae0417/packages/lexicon/src/validators/primitives.ts#L343) forces the value of `unknown` type to be object, but there is no such description in [the lexicon specification](https://atproto.com/specs/lexicon#unknown). From the specification, you can also refer to definitions with primitive types such as `string`, just like `ref`.

Is this a `unknown` specification or a validator-specific constraint?

**Comment by bnewbold** [comment, marked as answer]
This is a great question! I think this is a bit under-specified and we aren't sure about the details.

For example:

- can you distinguish between a null value and nullable value?
- are there limits on the data encoded in an unknown? eg, if it is an array, can it have a billion elements? I think this is mostly a generic data question, but it becomes more obvious with unknown fields
- can you have an array of `unknown`, with different types in each place?
- if an object, does it need to have `$type` included? I think not. can it have a `$type` though? do implementations need to distinguish between a generic object in an `unknown` and something like a blob (`$type: blob`)?
- what are the schema evolution rules? I think once something is `unknown` that probably can't change; but can something change from an object (or other type) to unknown? (I also think not)

Sorry to not have a more helpful answer. Pragmatically, right now I would try to stick to only using `object` in `unknown` fields, and we'll try to clarify this soon.

UPDATE: here is a tracking issue in the specs repo: https://github.com/bluesky-social/atproto-website/issues/319
